vmm/src/device_manager.rs

```diff
@@ -606,28 +607,38 @@ impl DeviceRelocation for AddressManager {
             PciBarRegionType::Memory32BitRegion | PciBarRegionType::Memory64BitRegion => {
                 // Update system allocator
                 if region_type == PciBarRegionType::Memory32BitRegion {
-                    self.allocator
-                        .lock()
-                        .unwrap()
-                        .free_mmio_hole_addresses(GuestAddress(old_base), len as GuestUsize);
+                    // Find the specific allocator that this BAR was allocated from and use it for new one
+                    for allocator in &self.pci_mmio32_allocators {
+                        let allocator_base = allocator.lock().unwrap().base();
+                        let allocator_end = allocator.lock().unwrap().end();
 
-                    self.allocator
-                        .lock()
-                        .unwrap()
-                        .allocate_mmio_hole_addresses(
-                            Some(GuestAddress(new_base)),
-                            len as GuestUsize,
-                            Some(len),
-                        )
-                        .ok_or_else(|| {
-                            io::Error::new(
-                                io::ErrorKind::Other,
-                                "failed allocating new 32 bits MMIO range",
-                            )
-                        })?;
+                        if old_base >= allocator_base.0 && old_base <= allocator_end.0 {
+                            allocator
+                                .lock()
+                                .unwrap()
+                                .free(GuestAddress(old_base), len as GuestUsize);
+
+                            allocator
+                                .lock()
+                                .unwrap()
+                                .allocate(
+                                    Some(GuestAddress(new_base)),
+                                    len as GuestUsize,
+                                    Some(std::cmp::min(len, 0x10)), // The minimum BAR alignment is 0x10
```

Comment:
nit: You can use the defined constant PVPANIC_DEVICE_MMIO_ALIGNMENT

Reply:
No, I don't think that is an appropriate use of that constant.

Reply:
I'm not sure amount the comment (perhaps it should say length rather than alignment) as this sets a minimum length. I think the PCI bar must be at least 4KiB aligned.

Reply:
Good point. After a closer reading, it looks like BARs must be naturally aligned, and the minimum BAR size is 0x10. With a 2-byte BAR `pv_panic` device is a bit out-of-spec (although it appears to work just fine anyway). For VFIO devices, there is an additional page-size alignment requirement that doesn't apply to emulated devices. I'm going to remove these two changes and focus on the memory allocator (but keep the change to the pv_panic device since that is sufficient to allow integration tests to pass).